### PR TITLE
add EntityVersion

### PR DESCRIPTION
**Pull request by MarcRaeder**
add RestAPI with Microsoft.EntityFramework

### CODE REVIEW

**Review by stefanluth — CHANGES_REQUESTED** [review]
Delete generated files and create a single `.gitignore` in the root dir :D

**Review by DevBastiS — CHANGES_REQUESTED** [review]
What Stefan said :D

**Review by DevBastiS — CHANGES_REQUESTED** [review]
Still 550 Files changed.
Add the bin folder to gitignore as well.

**Review by stefanluth — COMMENTED** [review]
Add EOFs :)

File: BookApi/BookApi/Controllers/BookController1.cs [hunk 1]
```diff
@@ -0,0 +1,49 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.EntityFrameworkCore;
+using BookApi.Models;
+using BookApi.Service;
```
Comment (stefanluth) [inline BookApi/BookApi/Controllers/BookController1.cs]: Some unused `usings`. Also, the file is named `BookController1.cs` :D Remove the `1`

File: BookApi/BookApi/Service/BookService.cs [hunk 2]
```diff
@@ -0,0 +1,33 @@
+using BookApi.Models;
+using NuGet.Protocol.Core.Types;
```
Comment (stefanluth) [inline BookApi/BookApi/Service/BookService.cs]: Unused `using`

File: BookApi/BookApi/Repository/BookRepository.cs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+            throw new Exception("Das Buch ist nicht vorhanden!");
+        }
+        bookToUpdate!.PageCount = newPageCount;
+        await bookContext.SaveChangesAsync();
+        
+
+        return bookToUpdate;
+    }
+    
+}
```
Comment (stefanluth) [inline BookApi/BookApi/Repository/BookRepository.cs:59]:
```suggestion
        await bookContext.SaveChangesAsync();

        return bookToUpdate;
    }

}
```

File: BookApi/BookApi/Service/BookService.cs [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    {
+        return await bookRepository.CreateBook(title, pageCount);
+    }
+    
+    public async Task<IList<Book>> DeleteAllBooks()
+    {
+        return await bookRepository.DeleteAllBooks();
+    }
+    
+    public async Task<Book> UpdatePageCountOfBook(string title, int newPageCount)
```
Comment (stefanluth) [inline BookApi/BookApi/Service/BookService.cs]:
```suggestion
    }

    public async Task<IList<Book>> DeleteAllBooks()
    {
        return await bookRepository.DeleteAllBooks();
    }

    public async Task<Book> UpdatePageCountOfBook(string title, int newPageCount)
```

File: BookApi/BookApi/Models/Book.cs [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+public class Book
+{
+    [Key]
+    [Required]
+    public Guid Id { get; set; }
+    
+    [Required]
+    public string Title { get; set; }
+    
+    public int? PageCount { get; set; }
```
Comment (stefanluth) [inline BookApi/BookApi/Models/Book.cs:16]:
```suggestion
    public Guid Id { get; set; }

    [Required]
    public string Title { get; set; }

    public int? PageCount { get; set; }
```

File: BookApi/BookApi/Controllers/BookController1.cs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+            this.bookService = bookService;
+        }
+
+        [HttpGet("getAllBooks")]
+        public async Task<IList<Book>> GetAllBooks()
+        {
+            return await bookService.GetAllBooks();
+        }
+        
+        [HttpPost("createBook")]
```
Comment (stefanluth) [inline BookApi/BookApi/Controllers/BookController1.cs]:
```suggestion
        }

        [HttpPost("createBook")]

```

**Review by stefanluth — CHANGES_REQUESTED** [review]
just code style and imports :)

File: BookApi/BookApi/Controllers/BookController1.cs [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.EntityFrameworkCore;
+using BookApi.Models;
+using BookApi.Service;
+
+namespace BookApi.Controllers
+{
+    [Route("")]
+    [ApiController]
+    public class BookController1 : ControllerBase
```
Comment (DevBastiS) [inline BookApi/BookApi/Controllers/BookController1.cs]:
```suggestion
    public class BookController : ControllerBase

```

File: BookApi/BookApi/Program.cs [hunk 8]
```diff
@@ -0,0 +1,32 @@
+using BookApi;
+using BookApi.Service;
+using BookApi.Models;
+using Microsoft.EntityFrameworkCore;
+
+var builder = WebApplication.CreateBuilder(args);
+
+// Add services to the container.
```
Comment (DevBastiS) [inline BookApi/BookApi/Program.cs]: Remove comments

File: BookApi/BookApi/Repository/BookRepository.cs [hunk 9]
```diff
@@ -0,0 +1,67 @@
+using BookApi.Models;
```
Comment (DevBastiS) [inline BookApi/BookApi/Repository/BookRepository.cs:1]: usings should be inside/below the namespace

File: BookApi/BookApi/Repository/BookRepository.cs [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+        try
+        {
+            await this.bookContext.Books.AddAsync(book);
+            await bookContext.SaveChangesAsync();
+
+            return book;
+        }
+        catch (Exception e)
+        {
+            Console.WriteLine(e);
```
Comment (DevBastiS) [inline BookApi/BookApi/Repository/BookRepository.cs]: If you are only writing the error to the console and then throwing it again you can work without the try catch.

File: BookApi/BookApi/Repository/BookRepository.cs [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+        this.bookContext.Books.RemoveRange(bookContext.Books);
+        await bookContext.SaveChangesAsync();
+
+        return await bookContext.Books.ToListAsync();
+    }
+    
+    public async Task<Book> UpdatePageCountOfBook(string title, int newPageCount)
+    {
+        var bookToUpdate = await bookContext.Books
+            .Where(book => book.Title == title)
```
Comment (DevBastiS) [inline BookApi/BookApi/Repository/BookRepository.cs]: You should work with primary key "id" here. The title could return more then one result.

File: BookApi/BookApi/Service/BookService.cs [hunk 12]
```diff
@@ -0,0 +1,33 @@
+using BookApi.Models;
+using NuGet.Protocol.Core.Types;
+
+namespace BookApi.Service;
+
+public class BookService
+{
+    private BookRepository bookRepository;
```
Comment (DevBastiS) [inline BookApi/BookApi/Service/BookService.cs:7]:
```suggestion
    private BookRepository BookRepository;
```

**Review by DevBastiS — CHANGES_REQUESTED** [review]

File: BookApi/BookApi/Repository/BookRepository.cs (hunk 9 above)
Comment (MarcRaeder) [reply BookApi/BookApi/Repository/BookRepository.cs:1]: if i place the usings under namespace i get errormessages

File: BookApi/BookApi/Controllers/BookController.cs [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+            return await bookService.CreateBook(title, pageCount);
+        }
+
+        [HttpDelete("deleteAllBooks")]
+        public async Task<IList<Book>> DeleteAllBooks()
+        {
+            return await bookService.DeleteAllBooks();
+        }
+
+        [HttpPut("updatePageCountOfBook")]
```
Comment (stefanluth) [inline BookApi/BookApi/Controllers/BookController.cs]: Use all lowercase and api routes and connect multiple words with hyphens :)
e.g.
`get-all-books`

File: BookApi/BookApi/Program.cs [hunk 14]
```diff
@@ -0,0 +1,32 @@
+using BookApi;
+using BookApi.Service;
+using BookApi.Models;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
```
Comment (stefanluth) [inline BookApi/BookApi/Program.cs]:
```suggestion
using Microsoft.EntityFrameworkCore;
```

VSCode shows these as unused imports :)

File: BookApi/BookApi/Service/BookService.cs [hunk 15]
```diff
@@ -0,0 +1,35 @@
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using BookApi.Models;
```
Comment (stefanluth) [inline BookApi/BookApi/Service/BookService.cs]:
```suggestion
using BookApi.Models;
```
VSCode shows these as unused imports :)

File: BookApi/BookApi/Repository/BookRepository.cs [hunk 16]
```diff
@@ -0,0 +1,63 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using BookApi.Models;
+using Microsoft.EntityFrameworkCore;
```
Comment (stefanluth) [inline BookApi/BookApi/Repository/BookRepository.cs]:
```suggestion
using BookApi.Models;
using Microsoft.EntityFrameworkCore;
```
VSCode shows these as unused imports :)

File: BookApi/BookApi/Models/Book.cs [hunk 17]
```diff
@@ -0,0 +1,24 @@
+using System;
```
Comment (stefanluth) [inline BookApi/BookApi/Models/Book.cs]:
```suggestion
```

File: BookApi/BookApi/Controllers/BookController.cs [hunk 18]
```diff
@@ -0,0 +1,46 @@
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Mvc;
+using BookApi.Models;
+using BookApi.Service;
```
Comment (stefanluth) [inline BookApi/BookApi/Controllers/BookController.cs]:
```suggestion
using Microsoft.AspNetCore.Mvc;
using BookApi.Models;
using BookApi.Service;

```

**Review by stefanluth — APPROVED** [review]

**Review by DevBastiS — APPROVED** [review]

**Pull request merged**